sdk/core/core-rest-pipeline/src/retryStrategies/failoverRetryStrategy.ts

```diff
@@ -0,0 +1,153 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { PipelineRequest } from "../interfaces";
+import { RetryInformation, RetryModifiers, RetryStrategy } from "./retryStrategy";
+import { DEFAULT_CLIENT_MAX_RETRY_INTERVAL, DEFAULT_CLIENT_RETRY_INTERVAL } from "../constants";
+import { DefaultRetryPolicyOptions } from "../policies/defaultRetryPolicy";
+
+/**
+ * A generator for failover hosts. Yields the next host to retry, or yields undefined if failover should be skipped.
+ */
+export type FailoverHostGenerator = (
+  retryState: RetryInformation,
+  options: DefaultRetryPolicyOptions
+) => Generator<FailoverHostState | undefined, void, RetryInformation>;
+
+/**
+ * A host for failover and its associated state.
+ */
+export declare interface FailoverHostState {
+  /**
+   * The host which this policy will try next.
+   */
+  host: string;
+  /**
+   * How many times this host has been tried.
+   */
+  retryCount: number;
+  /**
+   * The policy should wait until `retryAt` has elapsed to try this host.
+   */
+  retryAt?: Date;
+}
+declare interface FailoverHosts {
+  index: number;
+  hostStates: FailoverHostState[];
+}
+
+/**
+ * Default behavior for generating hosts for failover.
+ *
+ * If the request method is GET, HEAD, or OPTIONS, it cycles through `readHosts`. Otherwise, it cycles through `writeHosts`.
+ */
+export function defaultFailoverHostGenerator(
+  readHosts: string[] = [],
+  writeHosts: string[] = []
+): FailoverHostGenerator {
+  const readHostStateList: FailoverHosts = {
+    index: 0,
+    hostStates: readHosts.map((url) => {
+      const host = new URL(url).host;
+      return { host, retryCount: 0 };
+    }),
+  };
+  const writeHostStateList: FailoverHosts = {
+    index: 0,
+    hostStates: writeHosts.map((url) => {
+      const host = new URL(url).host;
+      return { host, retryCount: 0 };
+    }),
+  };
+  return function* (retryState, options) {
+    while (retryState) {
+      const response = retryState.response;
+      const retryDelayInMs = options.retryDelayInMs ?? DEFAULT_CLIENT_RETRY_INTERVAL;
+      const maxRetryDelayInMs = options.maxRetryDelayInMs ?? DEFAULT_CLIENT_MAX_RETRY_INTERVAL;
+      if (!response) {
+        return;
+      }
+
+      // Select correct list of hosts for this request
+      const isSafeMethod = ["GET", "HEAD", "OPTIONS"].includes(response.request.method);
+      const currentStateList = isSafeMethod ? readHostStateList : writeHostStateList;
+
+      if (currentStateList.hostStates.length === 0) {
+        yield undefined;
+        continue;
+      }
+      const hostState = currentStateList.hostStates[currentStateList.index];
+
+      const requestFireTime = Math.max(hostState.retryAt?.valueOf() ?? -Infinity, Date.now());
+
+      retryState = yield hostState;
+
+      const exponentialDelay = Math.min(
```

Comment:
maybe we should factor out the exponential delay logic from `exponentialRetryStrategy`? Especially since we do want to have jitter when doing failover, I think

Reply:
I think jitter makes sense here. I pulled that logic into a util function in 80238f5d6b1ba297deb5a2ab059ee3bd540a119a.